src/core/logging/Logger.hpp

```diff
@@ -0,0 +1,376 @@
+//
+// Created by loghin on 8/21/23.
+//
+
+#pragma once
+
+#include <CDS/Array>
+#include <CDS/memory/UniquePointer>
+
+#include <source_location>
+
+#include <lang/flag/FlagEnum.hpp>
+#include <lang/string/StringRef.hpp>
+
+namespace age {
+class Logger;
+
+namespace meta {
+enum class LogLevelFlagBits : cds::uint32 { Info = 1u << 0, Debug = 1u << 1, Error = 1u << 2, Warning = 1u << 3 };
+enum class LogOptionFlagBits : cds::uint32 {
+  OutputTerminalColour = 1u << 0,
+  InfoPrefix = 1u << 1,
+  SourceLocation = 1u << 8,
+  SourceLocationFile = 1u << 9,
+  SourceLocationFunction = 1u << 10,
+  SourceLocationLine = 1u << 11,
+  SourceLocationColumn = 1u << 12,
+  Timestamp = 1u << 16,
+  LoggerName = 1u << 24,
+  LogLevel = 1u << 25,
+  ThreadId = 1u << 26,
+};
+
+using LogLevelFlags = std::underlying_type_t<LogLevelFlagBits>;
+using LogOptionFlags = std::underlying_type_t<LogOptionFlagBits>;
+
+template <> struct FlagEnum<LogLevelFlagBits> : cds::meta::TrueType {
+  using FlagsType = LogLevelFlags;
+};
+
+template <> struct FlagEnum<LogOptionFlagBits> : cds::meta::TrueType {
+  using FlagsType = LogOptionFlags;
+};
+
+static constexpr auto const logLevelMask =
+    LogLevelFlagBits::Info | LogLevelFlagBits::Debug | LogLevelFlagBits::Warning | LogLevelFlagBits::Error;
+
+static constexpr auto const logOptionsMask = LogOptionFlagBits::OutputTerminalColour | LogOptionFlagBits::InfoPrefix
+    | LogOptionFlagBits::SourceLocation | LogOptionFlagBits::SourceLocationFile
+    | LogOptionFlagBits::SourceLocationFunction | LogOptionFlagBits::SourceLocationLine
+    | LogOptionFlagBits::SourceLocationColumn | LogOptionFlagBits::Timestamp | LogOptionFlagBits::LoggerName
+    | LogOptionFlagBits::LogLevel | LogOptionFlagBits::ThreadId;
+} // namespace meta
+
+class LoggerOutput {
+public:
+  static constexpr auto const allowInfo = meta::LogLevelFlagBits::Info;
+  static constexpr auto const allowWarning = meta::LogLevelFlagBits::Warning;
+  static constexpr auto const allowDebug = meta::LogLevelFlagBits::Debug;
+  static constexpr auto const allowError = meta::LogLevelFlagBits::Error;
+  static constexpr auto const allowAll = allowInfo | allowWarning | allowDebug | allowError;
+
+  using FilterFlags = meta::LogLevelFlags;
+  using FilterFlagBits = meta::LogLevelFlagBits;
+
+  explicit(false) LoggerOutput(std::ostream& out, FilterFlags filterFlags = allowAll) noexcept :
+      _pOutput(&out), _filter(filterFlags & mask) {}
+
+  LoggerOutput(std::ostream& out, FilterFlagBits filterLevel) noexcept :
+      _pOutput(&out), _filter(static_cast<FilterFlags>(filterLevel)) {}
+
+  [[nodiscard]] constexpr auto& output() noexcept { return *_pOutput; }
+  [[nodiscard]] constexpr auto const& output() const noexcept { return *_pOutput; }
+
+  [[nodiscard]] constexpr auto allows(meta::LogLevelFlags levels) const noexcept {
+    return (_filter & levels & mask) != 0u;
```

Comment:
Is & mask necessary?

Reply:
paranoia, and a double check doesn't hurt in this case
perhaps an assert is more fitting in this case